pkg/scheduler/framework/plugins/interpodaffinity/scoring.go

```diff
@@ -160,9 +160,14 @@ func (pl *InterPodAffinity) PreScore(
 		}
 	}
 
+	podInfo := framework.NewPodInfo(pod)
+	if podInfo.ParseError != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("parsing pod: %+v", err))
```

Comment:
/hold

Shouldn't we return `UnschedulableAndUnresolvable`?

`Error` implies internal error, but this is a user error.

Reply:
`UnschedulableAndUnresolvable` seems to only be parsed in Filter plugins, and the [error code definitions](https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/scheduler/framework/v1alpha1/interface.go#L59) imply that unexpected input could cause an error.

So `UnschedulableAndUnresolvable` could work for the Filter plugin, but score just returns anything that's not success as an error (https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/scheduler/core/generic_scheduler.go#L512-L516) (also, in 1.18 [this case is already returned as a `framework.Error`](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/scheduler/framework/plugins/interpodaffinity/filtering.go#L350) from the Filter plugin)

I think in the case of a syntax error, `Error` makes sense. `UnschedulableAndUnresolvable` sounds more related to cluster state that prevents the pod from scheduling (ie, it fails the filters and nothing can be preempted to fit it)

Reply:
What about doing UnschedulableAndUnresolvable for PreFilter, but just proceed with scheduling (ignore the badly formatted selectors) in the case of Score?

That doesn't solve the issue for 1.18 though. Unless we do what we are doing in 1.19.

The problem with using Error is that it messes up metrics. The metric should help identify internal errors, as opposed to user errors.

Reply:
We could do both checks (preferred and required constraints) in PreFilter and return `UnschedulableAndUnresolvable`. I think PreFilter is the right place to do any pod-level checks.

Reply:
I don't think this is going to mess up metrics too much, so I think it is fine if we are not accurate in 1.18.

Reply:
> What about doing UnschedulableAndUnresolvable for PreFilter, but just proceed with scheduling (ignore the badly formatted selectors) in the case of Score?


I think we should check the score so the user get a signal that there is something wrong with their pod spec rather than going undetected.

Reply:
wait, we are not going to get here anyways if there is any error because of the PreFilter check (unless the PreFilter is disabled).

Reply:
> just proceed with scheduling (ignore the badly formatted selectors) in the case of Score?

this was what caused one of the initial problems, a pod with bad formatting was ignored during scheduling and then broke scheduling for any subsequent pods that tried to check against the existing pod's affinity. that's actually how we found the bug, so we need to return some kind of error.

> The metric should help identify internal errors, as opposed to user errors.

Note this isn't a scheduling bug, it's a hack to patch an API validation bug. Ideally, just trying to create the pod would return a validation error from the api server but this is our workaround to accomplish the same effect of preventing the pod from running. 

I think that blurs the line between Error and Unschedulable but both are treated as !=Success so I'm fine with returning either. The only other thing is that it's already an Error from PreFilter in 1.18, so that would change too

Reply:
@alculquicondor pushed an update to change to `UnschedulableAndUnresolvable` in filter and score, wdyt?

Reply:
we should probably keep using Error in PreScore to emphasize that `UnschedulableAndUnresolvable` is not a PreScore expected status.